source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,282 @@
+          host, redis_discovery_session_->locality_lb_endpoint_);
+    }
+  }
+
+  priority_state_manager.updateClusterPrioritySet(
+      current_priority, std::move(priority_state_manager.priorityState()[current_priority].first),
+      hosts_added, hosts_removed, absl::nullopt);
+}
+
+// DnsDiscoveryResolveTarget
+RedisCluster::DnsDiscoveryResolveTarget::DnsDiscoveryResolveTarget(
+    RedisCluster& parent, const std::string& dns_address,
+    const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint,
+    const envoy::api::v2::endpoint::LbEndpoint& lb_endpoint)
+    : parent_(parent), dns_address_(dns_address), locality_lb_endpoint_(locality_lb_endpoint),
+      lb_endpoint_(lb_endpoint) {}
+
+RedisCluster::DnsDiscoveryResolveTarget::~DnsDiscoveryResolveTarget() {
+  if (active_query_) {
+    active_query_->cancel();
+  }
+}
+
+void RedisCluster::DnsDiscoveryResolveTarget::startResolve() {
+  ENVOY_LOG(trace, "starting async DNS resolution for {}", dns_address_);
+
+  active_query_ = parent_.dns_resolver_->resolve(
+      dns_address_, parent_.dns_lookup_family_,
+      [this](const std::list<Network::Address::InstanceConstSharedPtr>&& address_list) -> void {
+        active_query_ = nullptr;
+        ENVOY_LOG(trace, "async DNS resolution complete for {}", dns_address_);
+        parent_.redis_discovery_session_->registerDiscoveryAddress(address_list);
+        parent_.redis_discovery_session_->startResolve();
+      });
+}
+
+// RedisCluster
+RedisCluster::RedisDiscoverySession::RedisDiscoverySession(
+    Envoy::Extensions::Clusters::Redis::RedisCluster& parent,
+    NetworkFilters::Common::Redis::Client::ClientFactory& client_factory)
+    : parent_(parent),
+      resolve_timer_(parent.dispatcher_.createTimer([this]() -> void { startResolve(); })),
+      client_factory_(client_factory) {}
+
+inline Network::Address::InstanceConstSharedPtr
+ProcessCluster(const NetworkFilters::Common::Redis::RespValue& value) {
+  ASSERT(value.type() == NetworkFilters::Common::Redis::RespType::Array);
+  ASSERT(value.asArray().size() >= 2);
+  std::string address = value.asArray()[0].asString();
+  bool ipv6 = (address.find(":") != std::string::npos);
+  if (ipv6) {
+    return std::make_shared<Network::Address::Ipv6Instance>(address,
+                                                            value.asArray()[1].asInteger());
+  }
+  return std::make_shared<Network::Address::Ipv4Instance>(address, value.asArray()[1].asInteger());
+}
+
+RedisCluster::RedisDiscoverySession::~RedisDiscoverySession() {
+  if (current_request_) {
+    current_request_->cancel();
+    current_request_ = nullptr;
+  }
+
+  if (client_) {
+    client_->close();
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    // This should only happen after any active requests have been failed/cancelled.
```

Comment:
Is this really true? It looks to me like it should be possible to get a LocalClose event upon connection timeout of the underlying ClientImpl's connection with both current_request_ and client_ being non-null. I could be missing something, of course; all standard disclaimers apply.

Reply:
From debugging through `RedisClusterTest.RedisResolveFailure` it appears the failure case fires before anything else on connection timeout, which cleans up the request/connection. The tests mirror this. However, I'm not sure this actually happens in real life!

In any case I've broken out a helper function:

```
RedisCluster::RedisDiscoverySession::cancelRequestAndCloseClient() {
  if (current_request_) {
    current_request_->cancel();
    current_request_ = nullptr;
  }

  if (client_) {
    client_->close();
  }
}
```

Reply:
And I've unbroken this out, as it led to strange compile errors.